src/sync.rs

```diff
@@ -110,8 +114,10 @@ declare_id!(CondvarId);
 struct CondvarWaiter {
     /// The thread that is waiting on this variable.
     thread: ThreadId,
-    /// The mutex on which the thread is waiting.
-    mutex: MutexId,
+    /// The mutex or rwlock on which the thread is waiting.
+    lock: u32,
+    /// If the lock is shared or exclusive
+    shared: bool,
```

Comment:
I think you can encode both of these together and gain some extra type safety by using a new enum like
```rust
enum CondvarWaiterLock {
  Mutex(MutexId),
  RwLock(RwLockId),
}
```
Maybe the `RwLock` variant also needs a bool to determine if this is a read lock or write lock.